### PR TITLE
optimize compilation

### PR DESCRIPTION
**Pull request by vtjnash**
It seems like we weren't doing a great job with picking which methods to compile. This attempts to be much better at picking them. The new SSA IR compiler is also much more pessimistic in the runtime than the old compiler, which was better optimized to work efficiently. This helps optimize a few more of its code patterns and fixes a couple more missing method signature types.

### CODE REVIEW

**Comment by ViralBShah** [comment]
Review from nanosoldier? Is that the new way to do perf benchmarking?

File: base/compiler/ssair/passes.jl [hunk 1]
```diff
@@ -150,19 +153,25 @@ end
 Starting at `val` walk use-def chains to get all the leaves feeding into
 this val (pruning those leaves rules out by path conditions).
 """
-function walk_to_defs(compact, defssa, typeconstraint, visited_phinodes=Any[])
+function walk_to_defs(compact::IncrementalCompact, @nospecialize(defssa), @nospecialize(typeconstraint), visited_phinodes::Vector{Any}=Any[])
+    if !isa(defssa, Union{OldSSAValue, SSAValue, NewSSAValue}) || !isa(compact[defssa], PhiNode)
```
Comment (JeffBezanson) [inline base/compiler/ssair/passes.jl]: should use `AnySSAValue`

**Comment by vtjnash** [comment]
@nanosoldier `runbenchmarks(ALL, vs = ":master")`

**Review by JeffBezanson — APPROVED** [review]

**Comment by JeffBezanson** [comment]
Nice, this does seem to use less memory and CI seems to be quite happy with it so far.

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/27604#issuecomment-399520380) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/82049b44a6e1a6581e702b9f5c75d462c846cd93/2ca102c_vs_46a090b/report.md). cc @ararslan

**Comment by ararslan** [comment]
Looks pretty good overall. Going to rerun the scalar benchmarks to check, since those were the ones that showed the most regressions: @nanosoldier `runbenchmarks("scalar", vs=":master")`

**Comment by vtjnash** [comment]
Sounds good. There shouldn't be differences, so this run is just a sanity test, and a bit of a smoke screen test for nanosoldier itself.

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/27604#issuecomment-399616095) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/bed0e86a2d627577f2e18f119f18e384bcc9e2d6/4aa9796_vs_0eaa8b9/report.md). cc @ararslan

**Pull request merged**